### PR TITLE
[DET-2363] Port YOLOv3 to PEDL client.

### PR DESCRIPTION
**Pull request by AntonMu**
Port of the [ultralytics/yolov3](https://github.com/ultralytics/yolov3) PyTorch model to PEDL client.

Details of the model port are in the [README.md](README.md). 

Major changes from the original code were needed in cases when PEDL functions were unable to talk to each other. For instance, model did not have access to data and it was not possible to change optimizer parameters during training (depending on epochs).  

Overall, the porting was a lot smoother than with the old interface (IMHO).

### CODE REVIEW

File: Original_README.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+- **GCP** Deep Learning VM with $300 free credit offer: See our [GCP Quickstart Guide](https://github.com/ultralytics/yolov3/wiki/GCP-Quickstart) 
+- **Google Colab Notebook** with 12 hours of free GPU time: [Google Colab Notebook](https://colab.research.google.com/drive/1G8T-VFxQkjDe4idzN8F-hbIBqkkkQnxw)
+- **Docker Image** from https://hub.docker.com/r/ultralytics/yolov3. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov3/wiki/Docker-Quickstart) 
+# Citation
+
+[![DOI](https://zenodo.org/badge/146165888.svg)](https://zenodo.org/badge/latestdoi/146165888)
+
+# Contact
+
+**Issues should be raised directly in the repository.** For additional questions or comments please email Glenn Jocher at glenn.jocher@ultralytics.com or visit us at https://contact.ultralytics.com.
```
Comment (yoavz) [inline Original_README.md]: Probably just remove this file in this commit? It will still be around in the git commit history for reference.

File: README.md [hunk 2]
```diff
@@ -1,198 +1,75 @@
...
+Once downloaded, we need to edit `data/coco/train2017.txt` and `data/coco/val2017.txt` by replacing `../` by `data/` in front of each image file name. 
 
-`detect.py` runs inference on any sources:
+To make the data available to our script that runs with PEDL we use bind mounts in the experiment configuration:
 
-```bash
-python3 detect.py --source ...
+```
+config = {"bind_mounts": [{"host_path": "<abs_host_path_of_data>", "container_path": "data"}]}
 ```
```
Comment (yoavz) [inline README.md]: > to the PEDL master

this is incorrect, the data needs to be available to the PEDL agents, where the task environments are running.

Comment (yoavz) [inline README.md]: We're trying to standardize our best practices for data access: https://docs.google.com/document/d/1nwHQzIncy25RHItnslT4opa7xURVJivkWvm7RhykjJs/edit

Could investigate the possibility of uploading the data to a public GCS bucket and streaming from that directly? This would  allow us to remove this setup / bind mount configuration completely, which is error-prone and difficult.

File: pedl_train.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+def download_data() -> str:
+    # For a start we hard_code the dataset from bind_mounts
+    # Currently, PEDL Client does not support a download_data_fn
+    return "data"
+
+
```
Comment (yoavz) [inline pedl_train.py]: Can this function be omitted? 

If not, use an absolute path for the data instead of relative.

File: pedl_train.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+            return values
+
+        if self.last_epoch not in self.milestones:
+            return [group["lr"] for group in self.optimizer.param_groups]
+        return [
+            group["lr"] * self.gamma ** self.milestones[self.last_epoch]
+            for group in self.optimizer.param_groups
+        ]
+
+
```
Comment (yoavz) [inline pedl_train.py]: Can you factor out this scheduler implementation to a different file?

File: pedl_train.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+    # Hard code default arguments
+    opt = argparse.Namespace(
+        accumulate=4,
+        adam=True,
+        arc="default",
+        batch_size=16,
+        bucket="",
+        cache_images=False,
+        cfg="cfg/yolov3-spp.cfg",
+        data="data/coco2017.data",
```
Comment (yoavz) [inline pedl_train.py]: Use absolute paths for data

File: pedl_train.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+        model.arc = opt.arc  # attach yolo architecture
+        model.hyp = hyp  # attach hyperparameters to model
+
+        train_dataset = LazyModule.get()
+
+        # The model class weights depend on the dataset labels
+        model.class_weights = labels_to_class_weights(
+            train_dataset.labels, nc
+        )  # attach class weights
+
```
Comment (yoavz) [inline pedl_train.py]: It isn't clear where all these variables you're using are coming from? `hyp`, `nc`, `weights`? 

Are you creating them as globals in the `if __name__ == "__main__"` block and then somehow they're getting used here? I'm not really sure how that's even working -- regardless, it's a little confusing for reading through this example. I would recommend explicitly passing these configuration variables to the Trial class in a constructor:

```
def __init__(self, opt):
   self.opt

# later, initialize with the configuration
trial_class=YOLOv3Trial(opt)
```

That way a reader can at least trace how the configuration flows

File: pedl_train.py [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+                img_size,
+                batch_size,
+                augment=True,
+                hyp=hyp,  # augmentation hyperparameters
+                rect=opt.rect,  # rectangular training
+                cache_labels=True,
+                cache_images=opt.cache_images,
+                single_cls=opt.single_cls,
+            )
+            LazyModule.__instance = train_dataset
```
Comment (yoavz) [inline pedl_train.py]: you're using `train_path`, `img_size`, `batch_size`, bunch of variables, but it isn't clear where they're coming from -- similar to in the trial class. 

Refrain from using variables that are referenced outside of the scope of this class. you're assuming that `train_path`, `img_size`, `batch_size` are all defined in the global namespace before you call `get()` -- if you change the ordering later, it can be really tricky to debug. If you need configuration, pass it into the class directly .. e.g. via the `get()` function

File: pedl_train.py [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+
+
+class YOLOv3Trial(PyTorchTrial):
+    def build_model(self) -> nn.Module:
+
+        # Initialize model
+        model = Darknet(cfg, arc=opt.arc)  # .to(device)
+
+        # Fetch starting weights
+        attempt_download(weights)
```
Comment (yoavz) [inline pedl_train.py]: Any downloads should be in the `data_download` function. Since it isn't supported yet, you can leave it here, but add a todo to move it when supported.

File: pedl_train.py [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+        """
+        scheduler = WarmupMultiStepLR(
+            optimizer,
+            milestones=[round(opt.epochs * x) for x in [0.8, 0.9]],
+            gamma=0.1,
+            index=2,
+            early_value=0.1,
+        )
+
+        return LRScheduler(scheduler, step_every_epoch=True,)
```
Comment (yoavz) [inline pedl_train.py]: trailing comma

File: pedl_train.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+            "P",
+            "R",
+            "mAP_at_0.5",
+            "F1",
+            "Some_number_1",
+            "Some_number_2",
+            "Some_number_3",
+        ]
+
+        return dict(zip(keys[:4], results))
```
Comment (yoavz) [inline pedl_train.py]: Any reason you're only using the first 4 keys here? If it's unclear what the other numbers are, best to leave them out of `keys`.

File: pedl_train.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+    cfg = opt.cfg
+    data = opt.data
+    img_size, img_size_test = (
+        opt.img_size if len(opt.img_size) == 2 else opt.img_size * 2
+    )  # train, test sizes
+    images = opt.images
+    epochs = opt.epochs  # 500200 batches at bs 64, 117263 images = 273 epochs
+    batch_size = opt.batch_size
+    accumulate = opt.accumulate  # effective bs = batch_size * accumulate = 16 * 4 = 64
+    weights = opt.weights  # initial training weights
```
Comment (yoavz) [inline pedl_train.py]: is there a good reason to create a bunch of redundant variables? I would default to removing all of these variables and using `opt` directly throughout if not.

File: pedl_train.py [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+
+    # PEDL arguments
+    # We make some of the default hyperparameters and turn them into PEDL hyperparameters.
+
+    prebias = pedl.Constant(value=True, name="Prebias")
+    batch_size = pedl.Constant(value=16, name="batch_size")
+    init_lr = pedl.Constant(value=0.000579, name="init_lr")
+
+    # Once defined, we update the original hyperparameters with the PEDL values
+    opt.batch_size = batch_size
```
Comment (yoavz) [inline pedl_train.py]: Definitely do not define `batch_size = opt.batch_size` above if you're going to be overwriting it again here.

File: README.md (hunk 2 above)
Comment (AntonMu) [reply README.md]: Ok - I haven't worked with it. Would that automatically cache the data once it has been downloaded once? Brians PR seems a bit outdated now with the removal of task envs. If you have a little time, it would be good to talk about the best approach high level in this situation. Thanks!

File: pedl_train.py (hunk 3 above)
Comment (AntonMu) [reply pedl_train.py]: Yeah, it can be omitted - I kept it for future reference once download_data_fn is implemented in client. But happy to get rid of it for now.

File: pedl_train.py (hunk 4 above)
Comment (AntonMu) [reply pedl_train.py]: Sure

File: pedl_train.py (hunk 6 above)
Comment (AntonMu) [reply pedl_train.py]: I agree it is a bit confusing - I got them from the original script where they are all defined as globals. I was also surprised it worked but it was also nice bc I could basically use the original code and just make a few changes. 

Here is the part in the original code for hyp:

https://github.com/ultralytics/yolov3/blob/master/train.py#L25

And then there are the command line args in opt:

https://github.com/ultralytics/yolov3/blob/master/train.py#L394

File: pedl_train.py (hunk 7 above)
Comment (AntonMu) [reply pedl_train.py]: OK - again this was the way it was done originally and basically is a copy-paste of their code that worked "as-is" (and to my own surprise it worked as is). 

Overall it will boil down to the old questions what we want to optimize for:

 1. Minimize the number of code changes when porting an existing model or
 2. Have a good looking code in the end 

For 1, the good thing is that it would be relatively easy to give a complete walkthrough of the steps that were needed to port the model but at the price of ugly final code.

For 2, the good thing is nice looking code but the bad is that it will be harder to understand all the changes that were made and harder to compare to the original code.

File: pedl_train.py (hunk 10 above)
Comment (AntonMu) [reply pedl_train.py]: sure - forgot to remove that.

File: pedl_train.py (hunk 11 above)
Comment (AntonMu) [reply pedl_train.py]: Yeah - again this piece is copy-pasted from the orginal repo - I don't know why they did that. It is a bit weird I agree https://github.com/ultralytics/yolov3/blob/master/train.py#L56

**Comment by AntonMu** [comment]
Thanks for your comments. 

I agree that the original code isn't pretty and I wrote some comments above (quoted below) about two ways of think of it. Maybe that could be part of a broader discussion of best practices of using a popular github repo and porting it.

Now that we are moving towards needed less wrapper code, we could make a very clear division between existing code (maybe with block comments `#### This is original model code####` and then `### This is additional PEDL code needed ###` to separate their mess from our code). 

Or, we can also clean up and rewrite/restructure existing repos and publish a nicer version of it at the cost of less transparency. 

> OK - again this was the way it was done originally and basically is a copy-paste of their code that worked "as-is" (and to my own surprise it worked as is). 
> 
> Overall it will boil down to the old questions what we want to optimize for:
> 
>  1. Minimize the number of code changes when porting an existing model or
>  2. Have a good looking code in the end 
> 
> For 1, the good thing is that it would be relatively easy to give a complete walkthrough of the steps that were needed to port the model but at the price of ugly final code.
> 
> For 2, the good thing is nice looking code but the bad is that it will be harder to understand all the changes that were made and harder to compare to the original code.

File: README.md [hunk 13]
```diff
@@ -1,198 +1,75 @@
...
+```
 
-GPU |n| `--batch --accum` | img/s | epoch<br>time | epoch<br>cost
---- |--- |--- |--- |--- |---
-K80    |1| 32 x 2 | 11  | 175 min  | $0.58
-T4     |1<br>2| 32 x 2<br>64 x 1 | 41<br>61 | 48 min<br>32 min | $0.28<br>$0.36
-V100   |1<br>2| 32 x 2<br>64 x 1 | 122<br>**178** | 16 min<br>**11 min** | **$0.23**<br>$0.31
-2080Ti |1<br>2| 32 x 2<br>64 x 1 | 81<br>140 | 24 min<br>14 min | -<br>-
+#### Data Download
+To successfully use PEDL Client we need to make sure that all data is available both to your local environment and to the PEDL master. One way is to simply download the data to both your development local machine and to your PEDL master machine. Depending on your setup, you might be able to only download the data to one shared directory or simply point to a location that already has the data available. For simplicity, we just download the data to both machines by using the `bash get_coco2017.sh` script from within the `data` directory of this repo.
```
Comment (shiyuann) [inline README.md]: > To successfully use PEDL Client we need to make sure that all data is available both to your local environment and to the PEDL master.

1. PEDL Client -> PEDL.
2. both to your local environment and to the PEDL master -> all data is available both to the PEDL agents.

> Depending on your setup, you might be able to only download the data to one shared directory or simply point to a location that already has the data available.

What do you mean by `shared`?

File: README.md [hunk 14]
```diff
@@ -1,198 +1,75 @@
...
 ```
 
-- Image:  `--source file.jpg`
-- Video:  `--source file.mp4`
-- Directory:  `--source dir/`
-- Webcam:  `--source 0`
-- RTSP stream:  `--source rtsp://170.93.143.139/rtplive/470011e600ef003a004ee33696235daa`
-- HTTP stream:  `--source http://wmccpinetop.axiscam.net/mjpg/video.mjpg`
+#### Running it locally
+<!-- Question: How does it know which libraries need to be installed remotely? -->
```
Comment (shiyuann) [inline README.md]: What do you mean? Why not use startup-hook.sh here?

File: README.md [hunk 15]
```diff
@@ -1,198 +1,75 @@
...
+python train.py
+```
 
-**YOLOv3-tiny:** `python3 detect.py --cfg cfg/yolov3-tiny.cfg --weights yolov3-tiny.weights`  
-<img src="https://user-images.githubusercontent.com/26833433/64067834-51d5b500-cc2f-11e9-9357-c485b159a20b.jpg" width="500">
+### Running it with PEDL
 
-**YOLOv3-SPP:** `python3 detect.py --cfg cfg/yolov3-spp.cfg --weights yolov3-spp.weights`  
-<img src="https://user-images.githubusercontent.com/26833433/64067833-51d5b500-cc2f-11e9-8208-6fe197809131.jpg" width="500">
+We first install the PEDL command line interface. Then, we port over the original code to `pedl_train.py`. To port the code to PEDL, we need to implement a trial class, which we call `YOLOv3Trial` that inherits from PEDL's `PyTorchTrial` class. In this class, we define the model, the learning rate scheduler, the optimizer and the training and evaluation loop.
```
Comment (shiyuann) [inline README.md]: Make installing CLI a link.

File: pedl_train.py [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+            "metric": "mAP_at_0.5",
+            "max_steps": int(epochs * images * batch_size * accumulate / 100),
+            "smaller_is_better": False,
+        },
+        "max_restarts": 0,
+        "bind_mounts": [
+            {"host_path": "/home/anton/yolov3/data", "container_path": "data"}
+        ],
+        "optimizations": {"aggregation_frequency": accumulate},
+        "min_validation_period": int(
```
Comment (shiyuann) [inline pedl_train.py:396]: I think we could put how `max_steps` and `min_validation_period` are calculated in the README.md.

File: pedl_train.py [hunk 17]
```diff
@@ -0,0 +1,423 @@
...
+
+    config = {
+        "description": "yolov3_pytorch",
+        "searcher": {
+            "name": "single",
+            "metric": "mAP_at_0.5",
+            "max_steps": int(epochs * images * batch_size * accumulate / 100),
+            "smaller_is_better": False,
+        },
+        "max_restarts": 0,
```
Comment (shiyuann) [inline pedl_train.py]: This line seems to be unncessary.

File: README.md (hunk 14 above)
Comment (AntonMu) [reply README.md]: (this is commented out)

But I was curious how it works behind the scenes - I only installed all packaged on my local machine and then it seems like the packages were also available to pedl. This was a nice feature but I was wondering how it works.

File: README.md (hunk 13 above)
Comment (AntonMu) [reply README.md]: I changed the wording a bit in response to your comments - hope it is clearer now. 

Also, this section may get changed anyway bc of @yoavz comment on data access. 

By `shared`, I meant that for instance in my case, I develop on soak locally (outside of pedl) but also run PEDL on it. Thus, in this case, it is enough to download the data once locally and then use bind mounts to make it available in pedl.

File: README.md (hunk 14 above)
Comment (AntonMu) [reply README.md]: Looks like it was just a lucky coincidence that I didn't have to install any additional packages via startup hooks to make the code work. 

Currently, this is out-of-scope but in the future, it seems (@yoavz ) that we will support some sort of magic environment duplication on the pedl agensts. That is, all packages we used locally are automatically installed remotly (on pedl).

File: pedl_train.py (hunk 16 above)
Comment (AntonMu) [reply pedl_train.py:396]: ok

File: pedl_train.py (hunk 17 above)
Comment (AntonMu) [reply pedl_train.py]: sure - was just good for testing.

**Comment by AntonMu** [comment]
Implemented the suggestions. 

Only thing that is missing is the data access which I am not sure about - > see https://github.com/determined-ai/yolov3/pull/1#discussion_r379532875

File: pedl_train.py (hunk 5 above)
Comment (AntonMu) [reply pedl_train.py]: kept it for now as this is how it was used in the local development process.

File: pedl_train.py (hunk 6 above)
Comment (AntonMu) [reply pedl_train.py]: fixed.

File: pedl_train.py (hunk 7 above)
Comment (AntonMu) [reply pedl_train.py]: moved everything in separate `get_config` functions. The only question I have is if it impacts performance if we call this get_config for each batch training. Or if in this case it is better to use global variables.

File: pedl_train.py (hunk 11 above)
Comment (AntonMu) [reply pedl_train.py]: cleaned it up.

File: pedl_train.py (hunk 12 above)
Comment (AntonMu) [reply pedl_train.py]: removed.
